state/machine.go

```diff
@@ -2228,88 +2228,104 @@ func (m *Machine) SetUpgradeCharmProfile(appName, chURL string) error {
 		if life == Dead || life == Dying {
 			return nil, ErrDead
 		}
+		return m.SetUpgradeCharmProfileTxns(appName, chURL)
+	}
+	err := m.st.db().Run(buildTxn)
+	if err != nil {
+		return err
+	}
+	return nil
+}
 
-		charmURL, err := charm.ParseURL(chURL)
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
-		ch, err := m.st.Charm(charmURL)
-		if err != nil {
-			if !errors.IsNotFound(err) {
-				return nil, errors.Trace(err)
-			}
-		}
+// SetUpgradeCharmProfileTxns does the checks and creates the txns to
+// write an instanceCharmProfileDoc to trigger a profile change on a
+// machine.  AppName and Charm with an LXDProfile, adds or updates
+// the charm lxd profile on the machine. AppName and Charm without
+// an LXDProfile, removes the charm lxd profile from the machine if
+// was previously applied.  AppName without a Charm URL causes a
+// previously applied lxd profile for the charm to be removed from
+// the machine.
+func (m *Machine) SetUpgradeCharmProfileTxns(appName, chURL string) ([]txn.Op, error) {
+	// Check to see if the doc created in these txn already exists
+	// and has expected data.
+	err := m.verifyInstanceCharmProfileData(appName, chURL)
+	if err != nil {
+		return nil, err
```

Comment:
```suggestion
		return nil, errors.Trace(err)
```

Reply:
Turns out this would be not what we want.  The potential jujutxn.ErrNoOperations returned by verifyInstanceCharmProfileData() isn't handled correctly as a non fatal error if wrapped.  Other errors from verifyInstanceCharmProfileData() are wrapped in errors.Trace.